### PR TITLE
Dimension grain

### PR DESCRIPTION
**Pull request by nksocast**

### CODE REVIEW

File: lib/masamune/transform/postgres/deduplicate_dimension.psql.erb [hunk 1]
```diff
@@ -41,16 +42,16 @@ FROM (
     <%- target.insert_view_values.each do |value| -%>
     <%= value %><%= ',' %>
     <%- end -%>
-    start_at,
+    dimension_grain as start_at,
     CASE
-    WHEN <%= target.duplicate_value_conditions('w').join(' AND ') %> THEN
+    WHEN <%= target.duplicate_value_conditions('w',).join(' AND ') %>  THEN
```
Comment (mandrews) [inline lib/masamune/transform/postgres/deduplicate_dimension.psql.erb]: extra comma and space seem spurious

File: lib/masamune/transform/postgres/deduplicate_dimension.rb [hunk 2]
```diff
@@ -54,10 +54,11 @@ def insert_view_values(coalesce: false)
 
       def duplicate_value_conditions(window)
         [].tap do |result|
-          consolidated_columns.map do |_, column|
-            if column.null
-              result << "((LAG(#{column.name}) OVER #{window} = #{column.name}) OR (LAG(#{column.name}) OVER #{window} IS NULL AND #{column.name} IS NULL))"
-            else
+          # FIXME: add dimension_grain to dimension.rb, make grain as an attribute for a dimension
+          x = consolidated_columns
+          x[:dimension_grain] = OpenStruct.new(name: 'dimension_grain', null: false)
```
Comment (mandrews) [inline lib/masamune/transform/postgres/deduplicate_dimension.rb]: suggested refactor
```
 def consolidated_columns
   unreserved_columns.reject { |_, column| column.surrogate_key }.merge(dimension_grain: dimension_grain_column)
 end

 def dimension_grain_column
   OpenStruct.new(name: 'dimension_grain', null: false))
 end
```

File: lib/masamune/transform/postgres/deduplicate_dimension.psql.erb [hunk 3]
```diff
@@ -25,7 +25,8 @@ WITH consolidated AS (
     <%- target.insert_view_values(coalesce: true).each do |value| -%>
     <%= value %><%= ',' %>
     <%- end -%>
-    start_at
+    start_at,
+    date_trunc('hour',start_at) as dimension_grain
```
Comment (mandrews) [inline lib/masamune/transform/postgres/deduplicate_dimension.psql.erb]: `as` should be capitalized, e.g. `AS`

**Comment by nksocast** [comment]
feature pushed from another branch - https://github.com/socialcast/masamune/pull/164

**Pull request closed without merging**